torchdynamo/variables/functions.py

```diff
@@ -77,9 +77,10 @@ class UserFunctionVariable(BaseUserFunctionVariable):
 
     def __init__(self, fn, **kwargs):
         super(UserFunctionVariable, self).__init__(**kwargs)
+        # TODO(before land): verify if this is okay
         assert isinstance(
-            fn, types.FunctionType
-        ), f"expected FunctionType {typestr(fn)} {fn}"
+            fn, (types.FunctionType, types.MethodType)
```

Comment:
I don't think we can inline a method like this, because method has an object embedded in it `method.__self__`. To handle methods properly, we need to "unpack" it.  See `UserMethodVariable` below.

To handle this:
1) Pull out `method.__self__`.  You can construct a `VariableTracker` for `method.__self__` with something like:
```
self_var = VariableBuilder(tx, AttrSource(method_var.source, "__self__"))(method_var.value.__self__)
```
This will take an example value of `__self__` and build a Variable around it.  TorchDynamo will read the value dynamically from the `__self__` attribute.


2) Construct a `UserMethodVariable(method_var.value.__func__, self_var, ...)` (need to guard on this properly, but can ignore that for early testing)

3) Given a `UserMethodVariable`, torchdynamo should inline it automatically.

Reply:
to clarify, would the right place for the logic above be as an additional case in `UserMethodVariable.call_function`?

Reply:
You want to do this logic when you construct the Variable.  You want `method()` instances to map to UserMethodVariable (not UserFunctionVariable as they do in this PR)

Reply:
hmm, I might be missing something.  The reason I added the `types.MethodType` here is because `UserMethodVariable` inherits from `UserFunctionVariable`, and this code gets hit as a part of the constructor.  The current code (in misc.py) does create a `UserMethodVariable` already (the changes below, in this PR).

Reply:
current code in misc.py:

```
        elif isinstance(inner_fn, types.FunctionType):
            return variables.UserFunctionVariable(inner_fn, **options).call_function(
                tx, [self.objvar] + args, kwargs
            )
        elif isinstance(inner_fn, types.MethodType):
            return variables.UserMethodVariable(inner_fn, self.objvar, **options).call_function(
                tx, [self.objvar] + args, kwargs
            )
        else:
            unimplemented(f"non-function or method super: {inner_fn}")
```

This does create `UserMethodVariable`, it just doesn't do any of the special handling of `__self__`.  Should I add the extra self handling there?

Reply:
I think you just need to do `inner_fn.__func__` to get the underlying types.FunctionType.
```py
>>> class Foo:
...   def bar(self):
...     pass
... 
>>> type(Foo().bar)
<class 'method'>
>>> type(Foo().bar.__func__)
<class 'function'>
>>> 
```

Reply:
perfect, thx, that worked